man/man1/tcell.1
.TH TCELL 1 "16th April 2015" "PUPSP3 commands" "PUPSP3 commands"
.SH NAME
tcell \ licensing module (for arbitrary non PUPS codes) 
.br

.SH SYNOPSIS
.B tcell 
[
.B -pipe
.I command pipeline
] | [
.B -tee_file <tee_file_name>
][
.B -nice
.I cardinal
][
.B -argf
.I argument_file
][
.B -help
][
.B -usage
][
.B -growth_protection
.I retries
][
.B -slots
][
.B -slotinfo
]
[2> <log file>]
.br

.SH DESCRIPTION
.I tcell(1) 
is an application which runs the PUPS
.I securicor
security module on behalf of its parent process. If licensing conditions
are not met
.I tcell(1)
terminates the parent process. If licensing conditions are met, control is
passed back to the parent process.
.br

.SH OPTIONS
 
.B -help
invokes the on-line context help system. This
"man" page is displayed. The process then exits.
.br

.B -growth_protection
.I retrys
causes the system to retry an
.I sbrk(2)
call up to
.I retrys
times. This option permits the program to wait until the core that it requires
is available rather than simply exiting if
.I sbrk(2)
fails.
.br

.B -argf
.I argument_file
tells the filter to take its command line arguments from the specified
.I argument_file.
The data format of the argument file is:
.br

#  ... optional comment line ...
.B -argflag           [1]
.br
.I argument_parameter [1,1]
.br
.I argument_parameter [1,2]
.br

.I argument_parameter [1,n]
.br

# ... optional comment line ...
.br
.B -argflag           [2]
.br
.I argument_parameter [1,1]
.br
.I argument_parameter [1,2]
.br

.I argument_parameter [1,n]
.br
 
.B -nice
.I cardinal
determimes the
.I niceness
at which the filter is scheduled. In keeping with all UNIX user processes, nice
must be a cardinal between 0 and 20. The default niceness for
.I xtee(1)
is 4.
.br

.B -usage
displays program shortform usage.
.br

.B -version
displays program version number.
.br

.B -slots
displays the
.I pups(3)
libraries used in application in short form.
.br

.B -slotinfo
displays the
.I pups(3)
libraries used in application in long form giving version, source language,
programmer and copyright.

.SH LIBRARIES
.I xtee(1)
makes use of PUPS libraries [Portable UNIX Programming System], see
.I pups(3)
.br

.SH EXIT STATUS

255: error
.br
1:
.B -usage
or
.B -help
.br
0: no error
.br

.SH ENVIRONMENTAL VARIABLES
none.
.br

.SH COPYRIGHT
(C) Tumbling Dice Ltd, Gosforth.
.br

.SH PROGRAMMER
M.A. O'Neill.
.br


pupscore.src/tcell.c
/*-------------------------------------------------------------------------------------------------------
    Purpose: Tcell process -- attacks and destroys any (parent) process which does not hold the
             correct licence. 

    Author:  M.A. O'Neill
             Tumbling Dice Ltd
             Gosforth
             Newcastle upon Tyne
             NE3 4RT
             United Kingdom

    Version: 2.00
    Dated:   4th January 2022
    E-mail:  mao@tumblingdice.co.uk
-------------------------------------------------------------------------------------------------------*/


#include <me.h>
#include <utils.h>
#include <netlib.h>
#include <psrp.h>
#include <sys/file.h>
#include <sys/types.h>
#include <sys/mman.h>
#include <hipl_hdr.h>
#include <signal.h>
#include <unistd.h>
#include <string.h>
#include <errno.h>
#include <pheap.h>
#include <vstamp.h>
#include <casino.h>
#include <hash.h>
#include <stdlib.h>


/*---------------------*/
/* Application version */
/*---------------------*/

#define TCELL_VERSION    "2.00"


/*------------------------------------------------------------------------------------------------------
    Get application information for slot manager ...
------------------------------------------------------------------------------------------------------*/
/*---------------------------*/ 
/* Slot information function */
/*---------------------------*/ 

_PRIVATE void tcell_slot(int level)
{   (void)fprintf(stderr,"int dynamic app tcell %s: [ANSI C, PUPS MTD D]\n",TCELL_VERSION);
 
    if(level > 1)
    {  (void)fprintf(stderr,"(C) 2003-2022 Tumbling Dice\n");
       (void)fprintf(stderr,"Author: M.A. ONeill\n");
       (void)fprintf(stderr,"Unlicensed process destructor (PUPS format) (built %s %s)\n\n",__TIME__,__DATE__);
    }
    else
       (void)fprintf(stderr,"\n");

    (void)fflush(stderr);
}



/*----------------------------*/ 
/* Application usage function */
/*----------------------------*/ 

_PRIVATE void tcell_usage(void)

{   (void)fprintf(stderr,"tcell [>& <ASCII log file>]\n\n");
    (void)fprintf(stderr,"[>& <ASCII log file>]\n\n");

    (void)fprintf(stderr,"Signals\n\n");
    (void)fprintf(stderr,"SIGINIT SIGCHAN SIGPSRP: Process status [PSRP] request (protocol %4.2F)\n",PSRP_PROTOCOL_VERSION);
    (void)fprintf(stderr,"SICLIENT: tell client server is about to segment\n");

    #ifdef CRUI_SUPPORT
    (void)fprintf(stderr,"SIGCHECK SIGRESTART:      checkpoint and restart signals\n");
    #endif /* CRIU_SUPPORT */

    (void)fprintf(stderr,"SIGALIVE: check for existence of client on signal dispatch host\n");
    (void)fflush(stderr);
}


#ifdef SLOT
#include <slotman.h>
_EXTERN void (* SLOT)() __attribute__ ((aligned(16))) = tcell_slot;
_EXTERN void (* USE )() __attribute__ ((aligned(16))) = tcell_usage;
#endif /* SLOT */




/*-------------------------------------------------------------------------------------------------------
   Application build date ...
-------------------------------------------------------------------------------------------------------*/

_EXTERN char appl_build_time[SSIZE] = __TIME__;
_EXTERN char appl_build_date[SSIZE] = __DATE__;




/*-------------------------------------------------------------------------------------------------------
    Software I.D. tag (used if CKPT support enabled to discard stale dynamic
    checkpoint files) ...
-------------------------------------------------------------------------------------------------------*/

#define VTAG  4936
extern int appl_vtag = VTAG;




/*--------------------------------------------------------------------------------------------------------
    Exit function which terminates parent process (which has violated licence) ...
--------------------------------------------------------------------------------------------------------*/

_PRIVATE void violation(void)

{   (void)fprintf(stdout,"ERROR\n");
    (void)fflush(stdout);

    exit(255);
}



/*--------------------------------------------------------------------------------------------------------
    Main - decode command tail then interpolate using parameters supplied by user ...
--------------------------------------------------------------------------------------------------------*/

_PUBLIC int pups_main(int argc, char *argv[])

{   

    /*----------------------------------------------*/
    /* Register exit function (to terminate parent) */
    /*----------------------------------------------*/

    (void)pups_register_exit_f("violaton",(void *)&violation,(char *)NULL);


    /*---------------------------------------------------------------*/
    /* Do not allow PSRP clients to connect until we are initialised */
    /*---------------------------------------------------------------*/

    (void)psrp_ignore_requests();


    /*------------------------------------------*/
    /* Get standard items form the command tail */
    /*------------------------------------------*/
    /*------------------------------------------------------*/
    /* Error code 9999 will be emitted if securicor() fails */
    /* and this process (and its parent) will be terminated */
    /*------------------------------------------------------*/

    pups_std_init(TRUE,
                  &argc,
                  TCELL_VERSION,
                  "M.A. O'Neill",
                  "tcell",
                  "2022",
                  argv);


    /*--------------------------------------------*/
    /* We have license -- do not terminate parent */
    /*--------------------------------------------*/

    (void)pups_deregister_exit_f((void *)&violation);
    pups_exit(0);
}
